compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1017,7 +1018,43 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             return Ok(cycle_result);
         }
 
-        let (result, dep_node) = self.in_task(|this| this.evaluate_stack(&stack));
+        let (result, dep_node) = self.in_task(|this| {
+            let mut result = this.evaluate_stack(&stack)?;
+
+            // if evaluate failed, try to normalize the predicate and evaluate again.
+            if EvaluationResult::EvaluatedToErr == result
+                && fresh_trait_pred.has_projections()
+                && fresh_trait_pred.is_global()
+            {
+                let mut nested_obligations = Vec::new();
+                let predicate = try_normalize_with_depth_to(
+                    this,
+                    param_env,
+                    obligation.cause.clone(),
+                    obligation.recursion_depth + 1,
+                    fresh_trait_pred,
+                    &mut nested_obligations,
+                );
+                if predicate != fresh_trait_pred {
+                    let mut nested_result = EvaluationResult::EvaluatedToOk;
+                    for obligation in nested_obligations {
+                        nested_result = cmp::max(
+                            this.evaluate_predicate_recursively(stack.list(), obligation)?,
+                            nested_result,
+                        );
+                    }
+
+                    if nested_result.must_apply_modulo_regions() {
+                        let obligation = obligation.with(this.tcx(), predicate);
+                        result =
+                            this.evaluate_trait_predicate_recursively(stack.list(), obligation)?;
+                    }
+                }
```

Comment:
also "and" the `predicate` here together with the nested obligations from normalizing, a `must_apply_modulo_regions()` influencing behavior inside of the solver is pretty much always wrong.

Reply:
ah woops, `must_apply_modulo_regions()` (forgot what `modulo` means for a second, sorry) is actually alright. `considering_regions` is the variant which breaks even if just add an unused region parameter. Even if the nested result doesn't apply, you coudl still do `result = nested_result.max(this.evaluate_trait_predicate_recursively(stack.list(), obligation)?)` which is needed if the nested result could have region obligations.